### PR TITLE
Add keyboard to y&r studio.

### PR DESCRIPTION
**Pull request by jiaxin96**
Initially add keyboards (i.e., cool87: normal keyboard, no rgb; rainbow65: rgb, oled and encoder; sb65p: wireless split kb like redox-w; nlhhkb: normal keyboard)  and related tools (filp the font for oled and some things for raw hid feature to support rgb music visualization) for y&r studio.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
Please review the PR Checklist, and apply the suggested changes from it:
https://docs.qmk.fm/#/pr_checklist

**Review by qqprxzm — APPROVED** [review]

File: keyboards/yandrstudio/r65ble/info-via.json [hunk 1]
```diff
@@ -0,0 +1,157 @@
+{
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/info-via.json:1]: This does not belong in the qmk repo, and should be removed.

File: keyboards/yandrstudio/r65ble/jason_fount.c [hunk 2]
```diff
@@ -0,0 +1,137 @@
+#include "progmem.h"
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/jason_fount.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/yandrstudio/r65ble/keymaps/default/config.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/* OLED */
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

```suggestion
#pragma once
```

File: keyboards/yandrstudio/r65ble/keymaps/default/dev_oled/glcdfont.c [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#include "progmem.h"
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/dev_oled/glcdfont.c:1]: This appears to be the default font file and should be removed.

File: keyboards/yandrstudio/r65ble/keymaps/default/dev_oled/oled_driver.c [hunk 5]
```diff
@@ -0,0 +1,767 @@
+/*
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/dev_oled/oled_driver.c:1]: This file is a copy of the core files, and should be removed. 

If there are modifications to it, then those should be addressed in a separate PR, rather than copying the files and creating code fragmentation

File: keyboards/yandrstudio/r65ble/keymaps/default/dev_oled/oled_driver.h [hunk 6]
```diff
@@ -0,0 +1,313 @@
+/*
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/dev_oled/oled_driver.h:1]: Same as oled_driver.c

File: keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/kb_press_graph.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#include "kb_press_graph.h"
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/kb_press_graph.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/kb_press_graph.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
+#pragma once
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/kb_press_graph.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/my_pic_lib.h [hunk 9]
```diff
@@ -0,0 +1,136 @@
+#pragma once
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/my_pic_lib.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/wpm_graph.c [hunk 10]
```diff
@@ -0,0 +1,183 @@
+#include "wpm_graph.h"
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/wpm_graph.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/wpm_graph.h [hunk 11]
```diff
@@ -0,0 +1,151 @@
+#pragma once
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/graph_lib/wpm_graph.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/yandrstudio/r65ble/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,392 @@
+#include "r65ble.h"
+#include "dynamic_keymap.h"
+#include "raw_hid.h"
+#include "print.h"
+#include <string.h>
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/keymap.c:5]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/yandrstudio/r65ble/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+
+    // uprintf("%ud g_rgb_time scale8\n", scale16by8(g_rgb_timer, rgb_matrix_config.speed / 4));
+    if (clockwise) {
+        uint16_t keycode = dynamic_keymap_get_keycode(biton32(layer_state), 4, 3);
+        if (keycode >= MACRO00 && keycode <= MACRO15) {
+            dynamic_keymap_macro_send(keycode - MACRO00);
+        } else {
+            register_code16(keycode);
+            wait_ms(10);
+            unregister_code16(keycode);
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/keymap.c:85]:
```suggestion
            tap_code16(keycode);
```

File: keyboards/yandrstudio/r65ble/keymaps/default/rules.mk [hunk 14]
```diff
@@ -0,0 +1,11 @@
+# OLED_DRIVER_ENABLE = yes
+VIA_ENABLE = yes
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/rules.mk:2]: VIA should not be enabled in the default keymap, and should be in a separate keymap for via.

File: keyboards/yandrstudio/r65ble/keymaps/default/rules.mk [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+WPM_ENABLE = yes
+RGB_MATRIX_CUSTOM_USER = no
+CONSOLE_ENABLE = no
+TAP_DANCE_ENABLE = yes
+
+SRC += graph_lib/wpm_graph.c graph_lib/kb_press_graph.c dev_oled/oled_driver.c
+QUANTUM_LIB_SRC += i2c_master.c
+OPT_DEFS += -DOLED_DRIVER_ENABLE
+COMMON_VPATH += keyboards/yandrstudio/rainbow65/keymaps/test_oled/dev_oled
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/default/rules.mk:11]:
```suggestion
OLED_DRIVER_ENABLE = yes
```

File: keyboards/yandrstudio/r65ble/keymaps/test_oled/config.h [hunk 16]
```diff
@@ -0,0 +1,24 @@
+/* OLED */
```
Comment (drashna) [inline keyboards/yandrstudio/r65ble/keymaps/test_oled/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15


```suggestion
#pragma once
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/yandrstudio/r65ble/keymaps/default/keymap.c (hunk 13 above)
Comment (jiaxin96) [reply keyboards/yandrstudio/r65ble/keymaps/default/keymap.c:85]: I think the delay here is necessary, for example when a user uses some consumer controls, such as the volume control, which does not work properly on many systems without the delay.

Comment (fauxpark) [reply keyboards/yandrstudio/r65ble/keymaps/default/keymap.c:85]: You can set `TAP_CODE_DELAY 10` in your config.h then.

**Comment by jiaxin96** [comment]
Thanks for the suggestion from @drashna , I found there are still many problems in my code and will submit it after a second review.

**Comment by drashna** [comment]
Just a heads up, there is no need to close the PR.  

you can add additional commits to your branch and push those to github, and it will automatically update the pull request.  

Also, if you have the KBFirmware json, you can use this to generate code that is up to date: 
https://noroadsleft.github.io/kbf_qmk_converter/
